### ISSUE TITLE
Track technology and upgrade tallies over time. Expose on website.

### ISSUE THREAD

**Issue by lyricnz** [LukePrior/nbn-upgrade-map#265]
There's already a function to generate this - get_tech_and_upgrade_breakdown() - but it just emits output to the console.  It does take several minutes to run though, so shouldn't be included in every cycle.

```
All tech breakdown: 14652833
Counter({'FTTN': 4611720,
         'HFC': 2886497,
         'FTTP': 2793945,
         'FTTC': 1696950,
         'WIRELESS': 969169,
         'SATELLITE': 654873,
         'FTTB': 650671,
         'NULL': 389008})
All upgrade breakdown (excluding tech=FTTP): 11858888
Counter({'NULL_NA': 8339191,
         'FTTP_SA': 2030854,
         'UNKNOWN': 1351746,
         'FTTP_NA': 130383,
         'FW_NA': 4215,
         'FW_CT': 1767,
         'FTTN_CT': 552,
         'SAT_CT': 151,
         'FTTC_NA': 16,
         'FTTN_NA': 12,
         'FTTC_CT': 1})
```

**Comment by LukePrior** [comment]
Yeah this would be a very interesting statistic to watch over time, we could also partially backfill using the complete got history.

**Comment by lyricnz** [comment]
I'm experimenting with running this code on old checkouts of /results to generate historical data.

**Comment by lyricnz** [comment]
That's being somewhat stymied by the API changes a couple of months back (combined-suburbs.json etc). Will change to a pure geojson verison...

**Comment by lyricnz** [comment]
Running against git history - takes about 2 minutes per cycle (currently running every 7 days, starting now moving backwards)

So far

| date       |    FTTC |    FTTN |    FTTP |   WIRELESS |   NULL |   SATELLITE |     HFC |   FTTB |
|------------|---------|---------|---------|------------|--------|-------------|---------|--------|
| 2023-06-06 |  188161 | 1450549 |  402165 |      51390 |  48231 |       14352 |  384935 | 106112 |
| 2023-06-13 |  442519 | 3327931 |  826939 |      99443 | 125410 |       28110 |  842321 | 198282 |
| 2023-06-20 |  651212 | 3684540 | 1017752 |     185431 | 142307 |      101571 | 1137363 | 259840 |
| 2023-06-27 |  751954 | 3739463 | 1146575 |     222164 | 152780 |      133070 | 1238869 | 333044 |
| 2023-07-04 |  845328 | 3803819 | 1248432 |     267675 | 164681 |      172172 | 1390597 | 369067 |
| 2023-07-11 |  972679 | 3973614 | 1521368 |     334726 | 190493 |      235299 | 1578430 | 422578 |
| 2023-07-18 | 1155706 | 4250965 | 1834059 |     572103 | 254120 |      406338 | 1993516 | 491903 |
| 2023-07-25 | 1466512 | 4507919 | 2331807 |     804753 | 317568 |      524930 | 2566085 | 603075 |
| 2023-08-01 | 1704858 | 4674708 | 2683614 |     968538 | 383007 |      653705 | 2888947 | 651378 |
| 2023-08-08 | 1703028 | 4662432 | 2710729 |     968607 | 379691 |      653796 | 2888562 | 651736 |
| 2023-08-15 | 1700686 | 4659683 | 2715914 |     968541 | 379942 |      653865 | 2887755 | 650910 |
| 2023-08-22 | 1699955 | 4655818 | 2724126 |     968135 | 378928 |      653683 | 2888088 | 650786 |
| 2023-08-29 | 1698356 | 4657940 | 2727995 |     968403 | 378444 |      653975 | 2888486 | 650936 |
| 2023-09-05 | 1698277 | 4656768 | 2740027 |     969465 | 380229 |      654483 | 2889711 | 650787 |
| 2023-09-12 | 1697792 | 4611916 | 2785591 |     968243 | 386269 |      653989 | 2884607 | 650593 |
| 2023-09-19 | 1696950 | 4611721 | 2793974 |     969175 | 389007 |      654901 | 2886500 | 650672 |

| date       |   FTTP_SA |   NULL_NA |   UNKNOWN |   FTTP_NA |   FW_CT |   FTTN_CT |   FTTC_NA |   SAT_CT |   FTTN_NA |   FW_NA |   FTTC_CT |
|------------|-----------|-----------|-----------|-----------|---------|-----------|-----------|----------|-----------|---------|-----------|
| 2023-06-06 |    451082 |   1587551 |    170571 |     33927 |     337 |       257 |         2 |        1 |         1 |       1 |           |
| 2023-06-13 |   1070637 |   3521613 |    387502 |     83052 |     794 |       402 |         4 |        9 |         1 |       2 |           |
| 2023-06-20 |   1227185 |   4342624 |    500493 |     90576 |     893 |       471 |         5 |       13 |         2 |       2 |           |
| 2023-06-27 |   1280362 |   4645764 |    553159 |     90628 |     915 |       494 |         5 |       13 |         2 |       2 |           |
| 2023-07-04 |   1335730 |   4977284 |    607715 |     91137 |     938 |       513 |         5 |       13 |         2 |       2 |           |
| 2023-07-11 |   1402620 |   5514762 |    693051 |     95630 |    1137 |       522 |         5 |       15 |         3 |      74 |           |
| 2023-07-18 |   1508208 |   6585135 |    927248 |    101118 |    1294 |       546 |        10 |       20 |         8 |    1064 |           |
| 2023-07-25 |   1692635 |   7819799 |   1155645 |    118370 |    1782 |       584 |        17 |       90 |         9 |    1910 |         1 |
| 2023-08-01 |   1849033 |   8617697 |   1332511 |    120867 |    1794 |       585 |        18 |       95 |         9 |    2531 |         1 |
| 2023-08-08 |   1984878 |   8505559 |   1331781 |     80110 |    1785 |       568 |        19 |      103 |        19 |    3029 |         1 |
| 2023-08-15 |   1985668 |   8480853 |   1331047 |     97405 |    1790 |       562 |        17 |      129 |        18 |    3892 |         1 |
| 2023-08-22 |   1987903 |   8457693 |   1331567 |    111670 |    1791 |       559 |        16 |      141 |        13 |    4039 |         1 |
| 2023-08-29 |   1988844 |   8436536 |   1332266 |    132310 |    1792 |       557 |        16 |      156 |        13 |    4049 |         1 |
| 2023-09-05 |   2009861 |   8417893 |   1337316 |    127916 |    1782 |       556 |        16 |      158 |        13 |    4208 |         1 |
| 2023-09-12 |   2002659 |   8368338 |   1345948 |    129744 |    1773 |       552 |        16 |      150 |        13 |    4215 |         1 |
| 2023-09-19 |   2031663 |   8338278 |   1351789 |    130482 |    1767 |       552 |        16 |      151 |        12 |    4215 |         1 |

**Comment by lyricnz** [comment]
<img width="1412" alt="image" src="https://github.com/LukePrior/nbn-upgrade-map/assets/122371/5df881ba-06d7-4c46-b466-d05be8c23982">

**Comment by lyricnz** [comment]
Pushed PR and output https://github.com/LukePrior/nbn-upgrade-map/pull/266/files#diff-bbd7c3dbacf5aed3a6ba4d4847412e11a7b8098f831e6e660560a2823365d55f

**Comment by lyricnz** [comment]
FWIW project got to 100% address coverage on 31 July

**Comment by LukePrior** [comment]
Oh woah this looks great, might be a good idea to have an option to only include data from July 31st onwards.

**Comment by lyricnz** [comment]
Changed the processing to go old-to-new, starting on 23/5/23, and using 00:00 timestamp.

**Comment by LukePrior** [comment]
In terms of getting it onto the site, I might just quickly make a new page /stats or something and generate a few charts using Chart.js

**Comment by LukePrior** [comment]
Mockup I made with ChatGPT: https://lukeprior.github.io/nbn-upgrade-map/stats

**Comment by lyricnz** [comment]
I'll fix the code to make it update the same file, so it's rerunnable, without lots of little files.  Give me 30 mins

**Comment by LukePrior** [comment]
The breakdown.json file seems to have some issues at the start with dates and values.

**Comment by lyricnz** [comment]
Yup, I think I broke it while reformatting/merging it.  I'll just nuke it and rebuild it.

**Comment by LukePrior** [comment]
Merged, will update site then need to make GHA for automation each week.

**Comment by lyricnz** [comment]
I don't know if the git commands (in particular the checkout) will work within GHA.   We may need a different version of the code which just appends progress.json with the current checkout values.

**Comment by lyricnz** [comment]
Maybe a table on the stats page, for easier reading without lots of mouseover?

**Comment by lyricnz** [comment]
FYI "Helper function to sort tech data in descending order of values"
They already are (recent change)

**Comment by LukePrior** [comment]
https://nbn.lukeprior.com/stats
